### ISSUE TITLE
PHP error installing grpc from pecl -- BoringSSL does not build on Ubuntu 16.10

### ISSUE THREAD

**Issue by tbillington** [grpc/grpc#8670]
Getting error trying to `sudo pecl install grpc` from http://www.grpc.io/docs/quickstart/php.html on Ubuntu 16.10.

```bash
$ sudo pecl install grpc
downloading grpc-1.0.1.tgz ...
Starting to download grpc-1.0.1.tgz (1,910,227 bytes)
.....................................done: 1,910,227 bytes
884 source files, building
running: phpize
Configuring for:
PHP Api Version:         20151012
Zend Module Api No:      20151012
Zend Extension Api No:   320151012
building in /tmp/pear/temp/pear-build-roothPkMAS/grpc-1.0.1
running: /tmp/pear/temp/grpc/configure --with-php-config=/usr/bin/php-config
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for a sed that does not truncate output... /bin/sed
checking for cc... cc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether cc accepts -g... yes
checking for cc option to accept ISO C89... none needed
checking how to run the C preprocessor... cc -E
checking for icc... no
checking for suncc... no
checking whether cc understands -c and -o together... yes
checking for system library directory... lib
checking if compiler supports -R... no
checking if compiler supports -Wl,-rpath,... yes
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking target system type... x86_64-pc-linux-gnu
checking for PHP prefix... /usr
checking for PHP includes... -I/usr/include/php/20151012 -I/usr/include/php/20151012/main -I/usr/include/php/20151012/TSRM -I/usr/include/php/20151012/Zend -I/usr/include/php/20151012/ext -I/usr/include/php/20151012/ext/date/lib
checking for PHP extension directory... /usr/lib/php/20151012
checking for PHP installed headers prefix... /usr/include/php/20151012
checking if debug is enabled... no
checking if zts is enabled... no
checking for re2c... no
configure: WARNING: You will need re2c 0.13.4 or later if you want to regenerate PHP parsers.
checking for gawk... no
checking for nawk... nawk
checking if nawk is broken... no
checking whether to enable grpc support... yes, shared
checking how to print strings... printf
checking for a sed that does not truncate output... (cached) /bin/sed
checking for fgrep... /bin/grep -F
checking for ld used by cc... /usr/bin/ld
checking if the linker (/usr/bin/ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm -B
checking the name lister (/usr/bin/nm -B) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 1572864
checking how to convert x86_64-pc-linux-gnu file names to x86_64-pc-linux-gnu format... func_convert_file_noop
checking how to convert x86_64-pc-linux-gnu file names to toolchain format... func_convert_file_noop
checking for /usr/bin/ld option to reload object files... -r
checking for objdump... objdump
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for ar... ar
checking for archiver @FILE support... @
checking for strip... strip
checking for ranlib... ranlib
checking for gawk... (cached) nawk
checking command to parse /usr/bin/nm -B output from cc object... ok
checking for sysroot... no
checking for a working dd... /bin/dd
checking how to truncate binary pipes... /bin/dd bs=4096 count=1
checking for mt... mt
checking if mt is a manifest tool... no
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for dlfcn.h... yes
checking for objdir... .libs
checking if cc supports -fno-rtti -fno-exceptions... no
checking for cc option to produce PIC... -fPIC -DPIC
checking if cc PIC flag -fPIC -DPIC works... yes
checking if cc static flag -static works... yes
checking if cc supports -c -o file.o... yes
checking if cc supports -c -o file.o... (cached) yes
checking whether the cc linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... no
configure: creating ./config.status
config.status: creating config.h
config.status: executing libtool commands
running: make
........
/bin/bash /tmp/pear/temp/pear-build-roothPkMAS/grpc-1.0.1/libtool --mode=compile cc -Wall -Werror     -Wno-parentheses-equality -Wno-unused-value -std=c11     -fvisibility=hidden -DOPENSSL_NO_ASM -D_GNU_SOURCE -DWIN32_LEAN_AND_MEAN     -D_HAS_EXCEPTIONS=0 -DNOMINMAX -I. -I/tmp/pear/temp/grpc -DPHP_ATOM_INC -I/tmp/pear/temp/pear-build-roothPkMAS/grpc-1.0.1/include -I/tmp/pear/temp/pear-build-roothPkMAS/grpc-1.0.1/main -I/tmp/pear/temp/grpc -I/usr/include/php/20151012 -I/usr/include/php/20151012/main -I/usr/include/php/20151012/TSRM -I/usr/include/php/20151012/Zend -I/usr/include/php/20151012/ext -I/usr/include/php/20151012/ext/date/lib -I/tmp/pear/temp/grpc/include -I/tmp/pear/temp/grpc/src/php/ext/grpc -I/tmp/pear/temp/grpc/third_party/boringssl/include  -DHAVE_CONFIG_H  -g -O2   -c /tmp/pear/temp/grpc/third_party/boringssl/crypto/directory_posix.c -o third_party/boringssl/crypto/directory_posix.lo
libtool: compile:  cc -Wall -Werror -Wno-parentheses-equality -Wno-unused-value -std=c11 -fvisibility=hidden -DOPENSSL_NO_ASM -D_GNU_SOURCE -DWIN32_LEAN_AND_MEAN -D_HAS_EXCEPTIONS=0 -DNOMINMAX -I. -I/tmp/pear/temp/grpc -DPHP_ATOM_INC -I/tmp/pear/temp/pear-build-roothPkMAS/grpc-1.0.1/include -I/tmp/pear/temp/pear-build-roothPkMAS/grpc-1.0.1/main -I/tmp/pear/temp/grpc -I/usr/include/php/20151012 -I/usr/include/php/20151012/main -I/usr/include/php/20151012/TSRM -I/usr/include/php/20151012/Zend -I/usr/include/php/20151012/ext -I/usr/include/php/20151012/ext/date/lib -I/tmp/pear/temp/grpc/include -I/tmp/pear/temp/grpc/src/php/ext/grpc -I/tmp/pear/temp/grpc/third_party/boringssl/include -DHAVE_CONFIG_H -g -O2 -c /tmp/pear/temp/grpc/third_party/boringssl/crypto/directory_posix.c  -fPIC -DPIC -o third_party/boringssl/crypto/.libs/directory_posix.o
/tmp/pear/temp/grpc/third_party/boringssl/crypto/directory_posix.c: In function ‘OPENSSL_DIR_read’:
/tmp/pear/temp/grpc/third_party/boringssl/crypto/directory_posix.c:88:3: error: ‘readdir_r’ is deprecated [-Werror=deprecated-declarations]
   if (readdir_r((*ctx)->dir, &(*ctx)->dirent, &dirent) != 0 ||
   ^~
In file included from /tmp/pear/temp/grpc/third_party/boringssl/crypto/directory_posix.c:36:0:
/usr/include/dirent.h:183:12: note: declared here
 extern int readdir_r (DIR *__restrict __dirp,
            ^~~~~~~~~
/tmp/pear/temp/grpc/third_party/boringssl/crypto/directory_posix.c: At top level:
cc1: error: unrecognized command line option ‘-Wno-parentheses-equality’ [-Werror]
cc1: all warnings being treated as errors
Makefile:883: recipe for target 'third_party/boringssl/crypto/directory_posix.lo' failed
make: *** [third_party/boringssl/crypto/directory_posix.lo] Error 1
ERROR: `make' failed
```

**Comment by lhecker** [comment]
@hsaliak I upgraded `boringssl` to the newest commit in the chromium-stable branch. While running the tests for C the following tests fail:

```
FAILED: bins/opt/h2_compress_nosec_test resource_quota_server GRPC_POLL_STRATEGY=epoll
FAILED: bins/opt/h2_compress_nosec_test resource_quota_server GRPC_POLL_STRATEGY=poll
FAILED: bins/opt/h2_compress_nosec_test resource_quota_server GRPC_POLL_STRATEGY=poll-cv
FAILED: bins/opt/h2_compress_test filter_latency GRPC_POLL_STRATEGY=poll
FAILED: bins/opt/h2_compress_test resource_quota_server GRPC_POLL_STRATEGY=epoll
FAILED: bins/opt/h2_compress_test resource_quota_server GRPC_POLL_STRATEGY=poll
FAILED: bins/opt/h2_compress_test resource_quota_server GRPC_POLL_STRATEGY=poll-cv
FAILED: bins/opt/h2_full_nosec_test filter_latency GRPC_POLL_STRATEGY=epoll
FAILED: bins/opt/h2_load_reporting_test filter_latency GRPC_POLL_STRATEGY=poll
FAILED: bins/opt/handshake_client GRPC_POLL_STRATEGY=epoll
FAILED: bins/opt/handshake_client GRPC_POLL_STRATEGY=poll
FAILED: bins/opt/lb_policies_test GRPC_POLL_STRATEGY=epoll
FAILED: bins/opt/lb_policies_test GRPC_POLL_STRATEGY=poll
FAILED: bins/opt/lb_policies_test GRPC_POLL_STRATEGY=poll-cv
FAILED: bins/opt/sequential_connectivity_test GRPC_POLL_STRATEGY=epoll
FAILED: bins/opt/sequential_connectivity_test GRPC_POLL_STRATEGY=poll
FAILED: bins/opt/sequential_connectivity_test GRPC_POLL_STRATEGY=poll-cv
```

Are those errors to be expected? If yes I could submit a PR to solve this if you want to. 🙂

**Comment by hsaliak** [comment]
Bumped into the same issue on ubuntu 16.10.  I think the issue is not confined to PHP.
Removing PHP tag and bumping this up.

**Comment by emikulic** [comment]
I'm seeing the same on Debian unstable. This works around it:

`
diff --git a/Makefile b/Makefile
index eb0e216bc..2dac9138c 100644
--- a/Makefile
+++ b/Makefile
@@ -355,7 +355,7 @@ CXXFLAGS += -std=c++11
 else
 CXXFLAGS += -std=c++0x
 endif
-CPPFLAGS += -g -Wall -Wextra -Werror -Wno-long-long -Wno-unused-parameter
+CPPFLAGS += -g -Wall -Wextra -Werror -Wno-long-long -Wno-unused-parameter -Wno-error=deprecated-declarations
 LDFLAGS += -g
 
 CPPFLAGS += $(CPPFLAGS_$(CONFIG))
`

**Comment by stanley-cheung** [comment]
@nicolasnoble FYI, running `make` on a stock Ubuntu 16.10 instance fails when compiling boringssl.

```
[C]       Compiling third_party/boringssl/crypto/digest/digest.c
[C]       Compiling third_party/boringssl/crypto/digest/digests.c
[C]       Compiling third_party/boringssl/crypto/directory_posix.c
third_party/boringssl/crypto/directory_posix.c: In function ‘OPENSSL_DIR_read’:
third_party/boringssl/crypto/directory_posix.c:88:3: error: ‘readdir_r’ is deprecated [-Werror=deprecated-declarations]
   if (readdir_r((*ctx)->dir, &(*ctx)->dirent, &dirent) != 0 ||
   ^~
In file included from third_party/boringssl/crypto/directory_posix.c:36:0:
/usr/include/dirent.h:183:12: note: declared here
 extern int readdir_r (DIR *__restrict __dirp,
            ^~~~~~~~~
cc1: all warnings being treated as errors
Makefile:2267: recipe for target '/home/stanleycheung/grpc/objs/opt/third_party/boringssl/crypto/directory_posix.o' failed
```

I tried a couple newer commits on the boringssl commits and it didn't help as we seem to be set on a particular branch 2661.

**Comment by nicolasnoble** [comment]
Can we check if upgrading our boringssl version fixes this ?

**Comment by stanley-cheung** [comment]
Moving boringssl to this commit: ea2d217b9bc6c0bbbe8ded19b35584532c955ba0 seem to work

**Comment by hsaliak** [comment]
@nicolasnoble @stanley-cheung have we verified the fix, if not this might qualify as a release blocker?

**Comment by stanley-cheung** [comment]
So the fix #9340 is currently under review. @jboeuf is looking into an MSAN issue

**Comment by jboeuf** [comment]
Note that I'm at an offsite today and tomorrow. It's going to be hard for
me to resolve the msan issue during that time.
